cpu/atmega_common/periph/pwm.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2017 Víctor Ariño
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_atmega_common
+ * @ingroup     drivers_periph_pwm
+ * @{
+ *
+ * @file
+ * @brief       Low-level PWM driver implementation
+ *
+ * @author      Víctor Ariño <victor@lebrush.org>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "assert.h"
+#include "periph/pwm.h"
+#include "periph/gpio.h"
+
+#ifdef PWM_NUMOF
+
+#define WGM0 0
+#define WGM1 1
+#define WGM2 3
+#define COMA0 6
+#define COMA1 7
+
+static struct {
+    uint8_t CRA;
+    uint8_t CRB;
+    uint8_t res;
+} state[PWM_NUMOF];
+
+static inline unsigned get_prescaler(pwm_t dev, uint32_t *scale)
+{
+    uint16_t divmask = pwm_conf[dev].div;
+    uint32_t target = *scale;
+    uint8_t div, pre = 0;
+
+    for (div = 0; divmask > 0; div++) {
+        if (divmask & 0x01) {
+            pre++;
+            *scale = (uint32_t)(1 << div);
+            if (*scale >= target) {
+                break;
+            }
+        }
+        divmask >>= 1;
+    }
+
+    return pre;
+}
+
+uint32_t pwm_init(pwm_t dev, pwm_mode_t mode, uint32_t freq, uint16_t res)
+{
+    /* only left implemented, max resolution 256 */
+    assert(dev < PWM_NUMOF && mode == PWM_LEFT && res <= 256);
+    /* resolution != 256 only valid if ch0 not used */
+    assert(!(res != 256 && pwm_conf[dev].pin_ch[0] != GPIO_UNDEF));
+
+    /* disable PWM */
+    pwm_conf[dev].dev->CRA = 0x00;
+    pwm_conf[dev].dev->CRB = 0x00;
+    pwm_conf[dev].dev->OCR[0] = 0;
+    pwm_conf[dev].dev->OCR[1] = 0;
+
+    /* disable power reduction */
+    MEGA_PRR &= ~pwm_conf[dev].prr_bit;
```

Comment:
IMO this should already be in cpu_init, to turn off *all* peripherals by default to save energy!

Reply:
It's a good point, but no other platform is implemented as you suggest so far. I would rather keep it here by now.

Reply:
@lebrush We merged it today ( see https://github.com/RIOT-OS/RIOT/commit/914025973de120a011aa8f328205fb816a716a02 ). We are not using MEGA_PRR anymore.